Technically illegal syntax in ndimage generic filters

#3614 added a very cool generic filter to `cupyx.scipy.ndimage` supporting both `ReductionKernel` and `RawKernel`. However, when testing its compatibility with ROCm/HIP, I noticed there are two lines containing technically illegal syntax even in CUDA:
https://github.com/cupy/cupy/blob/a118905a7976079db8726001a3983d520847cb52/cupyx/scipy/ndimage/_filters_generic.py#L192
https://github.com/cupy/cupy/blob/a118905a7976079db8726001a3983d520847cb52/cupyx/scipy/ndimage/_filters_generic.py#L293
What these two lines do is to launch a kernel (i.e., the execution space specifier is `__global__`, to be precise) under the namespace `raw_kernel` inside another kernel, but without using any kernel launch syntax `<<< >>>` or applying dynamical parallelism with `-rdc` (which still requires the launch syntax), thus technically illegal as per CUDA Programming Guide. 

Given the fact that so far we saw no compiler warning/error and all tests passed, it seems NVRTC (or NVCC for that matter) did a poor job in catching this corner case, which is in fact a known issue on the host side: https://stackoverflow.com/questions/44285347/cuda-forgetting-kernel-launch-configuration-does-not-result-in-nvcc-compiler-wa, so I wouldn't be surprised if they fail to catch that on the device side too. 

However, hipRTC is able to detect it: 
```
E           AssertionError: Only cupy raises error
E           
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/testing/helper.py", line 37, in _call_func
E               result = impl(self, *args, **kw)
E             File "/home/leofang/dev/cupy_rocm3.5.0/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 348, in test_filter
E               return self._filter(xp, scp)
E             File "/home/leofang/dev/cupy_rocm3.5.0/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 90, in _filter
E               return filter(arr, *args, **kwargs)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupyx/scipy/ndimage/filters.py", line 1172, in generic_filter
E               weights_dtype=bool)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupyx/scipy/ndimage/_filters_core.py", line 139, in _call_kernel
E               kernel(*args)
E             File "cupy/core/_kernel.pyx", line 821, in cupy.core._kernel.ElementwiseKernel.__call__
E               kern = self._get_elementwise_kernel(dev_id, arginfos, type_map)
E             File "cupy/core/_kernel.pyx", line 846, in cupy.core._kernel.ElementwiseKernel._get_elementwise_kernel
E               kern = _get_elementwise_kernel(
E             File "cupy/_util.pyx", line 53, in cupy._util.memoize.decorator.ret
E               result = f(*args, **kwargs)
E             File "cupy/core/_kernel.pyx", line 639, in cupy.core._kernel._get_elementwise_kernel
E               return _get_simple_elementwise_kernel(
E             File "cupy/core/_kernel.pyx", line 37, in cupy.core._kernel._get_simple_elementwise_kernel
E               cpdef function.Function _get_simple_elementwise_kernel(
E             File "cupy/core/_kernel.pyx", line 60, in cupy.core._kernel._get_simple_elementwise_kernel
E               module = compile_with_cache(module_code, options)
E             File "cupy/core/core.pyx", line 1879, in cupy.core.core.compile_with_cache
E               return cuda.compile_with_cache(
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/cuda/compiler.py", line 382, in compile_with_cache
E               name_expressions, log_stream, cache_in_memory)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/cuda/compiler.py", line 758, in _compile_with_cache_hip
E               log_stream, cache_in_memory)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/cuda/compiler.py", line 221, in compile_using_nvrtc
E               name_expressions, log_stream, jitify)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/cuda/compiler.py", line 204, in _compile
E               ptx, mapping = prog.compile(options, log_stream)
E             File "/home/leofang/dev/cupy_rocm3.5.0/cupy/cuda/compiler.py", line 589, in compile
E               'nvrtc' if not runtime.is_hip else 'hiprtc')
E           /tmp/comgr-bcd757/input/CompileSource:5350:17: error: reference to __global__ function 'rms' in __global__ function
E               raw_kernel::rms(values, 9, &val_out);
E                           ^
E           /tmp/comgr-bcd757/input/CompileSource:5295:6: note: 'rms' declared here
E           void rms(const double* x, int filter_size, double* y) {
E                ^
E           1 error generated when compiling for gfx906.
E           Error: Failed to compile opencl source (from CL or HIP source to LLVM IR).
```

@coderforlife, I discussed offline with @emcastillo and we think your intention is to use `ReductionKernel`/`RawKernel` as a `__device__` function here, are we understanding it correctly? If so, we have a simple fix that I will test on AMD GPUs later.

--- a/cupy/core/core.pyx
+++ b/cupy/core/core.pyx
@@ -1735,6 +1735,9 @@ cdef list _cupy_header_list = [
     'cupy/carray.cuh',
     'cupy/atomics.cuh',
 ]
+if runtime._is_hip_environment:
+    _cupy_header_list.append('cupy/math_constants.h')
+
 cdef str _cupy_header = ''.join(
     ['#include <%s>\n' % i for i in _cupy_header_list])
 

--- a/cupy/core/include/cupy/math_constants.h
+++ b/cupy/core/include/cupy/math_constants.h
@@ -1,0 +1,110 @@
+#ifndef CUPY_MATH_CONSTANTS_H
+#define CUPY_MATH_CONSTANTS_H
+
+#ifdef __CUDA_ARCH__
+
+#include <math_constants.h>
+
+#elif (defined(__HIP_DEVICE_COMPILE__) || defined(CUPY_USE_HIP))
+
+/* single precision constants */
+#define CUDART_INF_F            __int_as_float(0x7f800000)
+#define CUDART_NAN_F            __int_as_float(0x7fffffff)
+#define CUDART_MIN_DENORM_F     __int_as_float(0x00000001)
+#define CUDART_MAX_NORMAL_F     __int_as_float(0x7f7fffff)
+#define CUDART_NEG_ZERO_F       __int_as_float(0x80000000)
+#define CUDART_ZERO_F           0.0f
+#define CUDART_ONE_F            1.0f
+#define CUDART_SQRT_HALF_F      0.707106781f
+#define CUDART_SQRT_HALF_HI_F   0.707106781f
+#define CUDART_SQRT_HALF_LO_F   1.210161749e-08f
+#define CUDART_SQRT_TWO_F       1.414213562f
+#define CUDART_THIRD_F          0.333333333f
+#define CUDART_PIO4_F           0.785398163f
+#define CUDART_PIO2_F           1.570796327f
+#define CUDART_3PIO4_F          2.356194490f
+#define CUDART_2_OVER_PI_F      0.636619772f
+#define CUDART_SQRT_2_OVER_PI_F 0.797884561f
+#define CUDART_PI_F             3.141592654f
+#define CUDART_L2E_F            1.442695041f
+#define CUDART_L2T_F            3.321928094f
+#define CUDART_LG2_F            0.301029996f
+#define CUDART_LGE_F            0.434294482f
+#define CUDART_LN2_F            0.693147181f
+#define CUDART_LNT_F            2.302585093f 
+#define CUDART_LNPI_F           1.144729886f
+#define CUDART_TWO_TO_M126_F    1.175494351e-38f
+#define CUDART_TWO_TO_126_F     8.507059173e37f
+#define CUDART_NORM_HUGE_F      3.402823466e38f
+#define CUDART_TWO_TO_23_F      8388608.0f
+#define CUDART_TWO_TO_24_F      16777216.0f
+#define CUDART_TWO_TO_31_F      2147483648.0f
+#define CUDART_TWO_TO_32_F      4294967296.0f
+#define CUDART_REMQUO_BITS_F    3
+#define CUDART_REMQUO_MASK_F    (~((~0)<<CUDART_REMQUO_BITS_F))
+#define CUDART_TRIG_PLOSS_F     105615.0f
+
+/* double precision constants */
+#define CUDART_INF              __longlong_as_double(0x7ff0000000000000ULL)
+#define CUDART_NAN              __longlong_as_double(0xfff8000000000000ULL)
+#define CUDART_NEG_ZERO         __longlong_as_double(0x8000000000000000ULL)
+#define CUDART_MIN_DENORM       __longlong_as_double(0x0000000000000001ULL)
+#define CUDART_ZERO             0.0
+#define CUDART_ONE              1.0
+#define CUDART_SQRT_TWO         1.4142135623730951e+0
+#define CUDART_SQRT_HALF        7.0710678118654757e-1
+#define CUDART_SQRT_HALF_HI     7.0710678118654757e-1
+#define CUDART_SQRT_HALF_LO   (-4.8336466567264567e-17)
+#define CUDART_THIRD            3.3333333333333333e-1
+#define CUDART_TWOTHIRD         6.6666666666666667e-1
+#define CUDART_PIO4             7.8539816339744828e-1
+#define CUDART_PIO4_HI          7.8539816339744828e-1
+#define CUDART_PIO4_LO          3.0616169978683830e-17
+#define CUDART_PIO2             1.5707963267948966e+0
+#define CUDART_PIO2_HI          1.5707963267948966e+0
+#define CUDART_PIO2_LO          6.1232339957367660e-17
+#define CUDART_3PIO4            2.3561944901923448e+0
+#define CUDART_2_OVER_PI        6.3661977236758138e-1
+#define CUDART_PI               3.1415926535897931e+0
+#define CUDART_PI_HI            3.1415926535897931e+0
+#define CUDART_PI_LO            1.2246467991473532e-16
+#define CUDART_SQRT_2PI         2.5066282746310007e+0
+#define CUDART_SQRT_2PI_HI      2.5066282746310007e+0
+#define CUDART_SQRT_2PI_LO    (-1.8328579980459167e-16)
+#define CUDART_SQRT_PIO2        1.2533141373155003e+0
+#define CUDART_SQRT_PIO2_HI     1.2533141373155003e+0
+#define CUDART_SQRT_PIO2_LO   (-9.1642899902295834e-17)
+#define CUDART_SQRT_2OPI        7.9788456080286536e-1
+#define CUDART_L2E              1.4426950408889634e+0
+#define CUDART_L2E_HI           1.4426950408889634e+0
+#define CUDART_L2E_LO           2.0355273740931033e-17
+#define CUDART_L2T              3.3219280948873622e+0
+#define CUDART_LG2              3.0102999566398120e-1
+#define CUDART_LG2_HI           3.0102999566398120e-1
+#define CUDART_LG2_LO         (-2.8037281277851704e-18)
+#define CUDART_LGE              4.3429448190325182e-1
+#define CUDART_LGE_HI           4.3429448190325182e-1
+#define CUDART_LGE_LO           1.09831965021676510e-17
+#define CUDART_LN2              6.9314718055994529e-1
+#define CUDART_LN2_HI           6.9314718055994529e-1
+#define CUDART_LN2_LO           2.3190468138462996e-17
+#define CUDART_LNT              2.3025850929940459e+0
+#define CUDART_LNT_HI           2.3025850929940459e+0
+#define CUDART_LNT_LO         (-2.1707562233822494e-16)
+#define CUDART_LNPI             1.1447298858494002e+0
+#define CUDART_LN2_X_1024       7.0978271289338397e+2
+#define CUDART_LN2_X_1025       7.1047586007394398e+2
+#define CUDART_LN2_X_1075       7.4513321910194122e+2
+#define CUDART_LG2_X_1024       3.0825471555991675e+2
+#define CUDART_LG2_X_1075       3.2360724533877976e+2
+#define CUDART_TWO_TO_23        8388608.0
+#define CUDART_TWO_TO_52        4503599627370496.0
+#define CUDART_TWO_TO_53        9007199254740992.0
+#define CUDART_TWO_TO_54        18014398509481984.0
+#define CUDART_TWO_TO_M54       5.5511151231257827e-17
+#define CUDART_TWO_TO_M1022     2.22507385850720140e-308
+#define CUDART_TRIG_PLOSS       2147483648.0
+#define CUDART_DBL2INT_CVT      6755399441055744.0
+
+#endif
+#endif  // CUPY_MATH_CONSTANTS_H

--- a/cupyx/scipy/ndimage/_filters_core.py
+++ b/cupyx/scipy/ndimage/_filters_core.py
@@ -143,7 +143,10 @@ def _call_kernel(kernel, input, weights, output, structure=None,
     return output
 
 
-math_constants_preamble = "#include <cupy/math_constants.h>\n"
+math_constants_preamble = r'''
+// workaround for HIP: line begins with #include
+#include <cupy/math_constants.h>
+'''
 
 
 _CAST_FUNCTION = """

--- a/cupyx/scipy/ndimage/_filters_core.py
+++ b/cupyx/scipy/ndimage/_filters_core.py
@@ -143,7 +143,7 @@ def _call_kernel(kernel, input, weights, output, structure=None,
     return output
 
 
-math_constants_preamble = "#include <math_constants.h>\n"
+math_constants_preamble = "#include <cupy/math_constants.h>\n"
 
 
 _CAST_FUNCTION = """

--- a/cupyx/scipy/ndimage/_filters_generic.py
+++ b/cupyx/scipy/ndimage/_filters_generic.py
@@ -196,7 +196,12 @@ def _get_generic_filter_raw(rk, filter_size, mode, wshape, offsets, cval,
         'generic_{}_{}'.format(filter_size, rk.name),
         setup, 'values[iv++] = cast<double>({value});', sub_call,
         mode, wshape, int_type, offsets, cval,
-        preamble='namespace raw_kernel {{\n{}\n}}'.format(rk.code),
+        preamble='namespace raw_kernel {{\n{}\n}}'.format(
+            # Users can test RawKernel independently, but when passed to here
+            # it must be used as a device function here. In fact, RawKernel
+            # wouldn't compile if code only contains device functions, so this
+            # is necessary.
+            rk.code.replace('__global__', '__device__')),
         options=rk.options)
 
 
@@ -301,6 +306,11 @@ void {name}(const byte* input, byte* output, const idx_t* x) {{
 }}'''.format(n_lines=n_lines, length=length, in_length=in_length, start=start,
              in_ctype=in_ctype, out_ctype=out_ctype, int_type=int_type,
              boundary_early=boundary_early, boundary=boundary,
-             name=name, rk_name=rk.name, rk_code=rk.code,
+             name=name, rk_name=rk.name,
+             # Users can test RawKernel independently, but when passed to here
+             # it must be used as a device function here. In fact, RawKernel
+             # wouldn't compile if code only contains device functions, so this
+             # is necessary.
+             rk_code=rk.code.replace('__global__', '__device__'),
              CAST=_filters_core._CAST_FUNCTION)
     return cupy.RawKernel(code, name, ('--std=c++11',) + rk.options)

--- a/cupyx/scipy/ndimage/_interp_kernels.py
+++ b/cupyx/scipy/ndimage/_interp_kernels.py
@@ -5,7 +5,7 @@ import cupy.core.internal
 
 from cupyx.scipy.ndimage import _util
 
-math_constants_preamble = "#include <math_constants.h>\n"
+math_constants_preamble = "#include <cupy/math_constants.h>\n"
 
 
 def _get_coord_map(ndim):

--- a/cupyx/scipy/special/_convex_analysis.py
+++ b/cupyx/scipy/special/_convex_analysis.py
@@ -2,7 +2,7 @@ from cupy import core
 
 
 _float_preamble = '''
-#include <math_constants.h>
+#include <cupy/math_constants.h>
 
 double __device__ entr(double x) {
     if (isnan(x)) {

